src/poligrain/spatial.py

```diff
@@ -477,3 +478,128 @@ def get_grid_time_series_at_intersections(grid_data, intersect_weights):
+    )
+
+    coords_cml_b = np.hstack(
+        [ds_cmls.site_1_y.data.reshape(-1, 1), ds_cmls.site_1_x.data.reshape(-1, 1)]
+    )
+
+    coords_gauge = np.hstack(
+        [ds_gauges.y.data.reshape(-1, 1), ds_gauges.x.data.reshape(-1, 1)]
+    )
+
+    cml_lengths = ds_cmls.length.data / 2
+
+    # calculate CML midpoints by using the average of site a and b
+    coords_cml = np.hstack(
+        [
+            ((coords_cml_a[:, 0] + coords_cml_b[:, 0]) / 2).reshape(-1, 1),
+            ((coords_cml_a[:, 1] + coords_cml_b[:, 1]) / 2).reshape(-1, 1),
+        ]
+    )
+
+    # extend max_distance to include diagonal length
+    max_distance_add = np.sqrt(2 * max_distance**2)
+
+    # array for storing indices of gauges close to CML, -1 represent no gauge
+    list_gauges = np.ones([ds_cmls.cml_id.size, n_closest]).astype(int) * (-1)
+
+    # array for storing distances from CML to close gauges
+    cml_gauge_dist = np.zeros([ds_cmls.cml_id.size, n_closest]) * np.nan
+
+    # create KDTree object for all gauges
+    kd_tree2 = KDTree(coords_gauge)
+
+    for i in range(len(ds_cmls.cml_id)):
+        # create KDTree object for midpoint of CML_i
+        kd_tree1 = KDTree([coords_cml[i]])  # select cml i
+
+        # find points within
+        sdm = kd_tree1.sparse_distance_matrix(
+            kd_tree2, cml_lengths[i] + max_distance_add
+        )
+
+        # to sparese distance matrix
+        sdm_coo = sdm.tocoo()
+
+        # get indices of gauges within max_distance
+        non_zero_cols = sdm_coo.col
+
+        # create line object for this CML
+        line = LineString([coords_cml_a[i], coords_cml_b[i]])
+
+        # calculate distance from CML to close gauges
+        d_to_gauges = line.distance(
+            [Point([coords_gauge[j, 0], coords_gauge[j, 1]]) for j in non_zero_cols]
+        )
+
+        # sort distances and get corresponding indices
+        sorted_indices = np.argsort(d_to_gauges)
+
+        # get distance of gauges close to CML
+        d_to_closest_gauges = d_to_gauges[sorted_indices[0:n_closest]]
+
+        # get indices of gauges close to CML
+        ind_closest_gauges = non_zero_cols[sorted_indices[0:n_closest]]
+
+        # remove points outside max_distance
+        d_to_closest_gauges[d_to_closest_gauges > max_distance] = np.nan
+        ind_closest_gauges = ind_closest_gauges[~np.isnan(d_to_closest_gauges)]
+
+        # store results for this CML
+        list_gauges[i, : ind_closest_gauges.size] = ind_closest_gauges
+
+        cml_gauge_dist[i, : d_to_closest_gauges.size] = d_to_closest_gauges
+
+    # create xarray object showing name and distance from cml to nearest points
+    return xr.Dataset(
+        data_vars={
+            "distance": (("cml_id", "n_closest"), cml_gauge_dist),
+            "id_neighbor": (
+                ("cml_id", "n_closest"),
+                np.where(list_gauges == -1, np.nan, ds_gauges.id.data[list_gauges]),
```

Comment:
I understand how you do that with the -1 in `list_gauges`, it is just hard to understand everything good enough to be sure that things work correctly. Maybe we can discuss that online at some point before you finish this PR.

Reply:
good point, the code is hard to read. I will see if I can make it more readable without compromising the speed and then we can talk through it online.

Reply:
Note that I just had the same problem in my PR. The problem is that the `KDTree.query()` returns indices which are out of bounds for the case where distance is larger than `max_distance`, right?

I have a WIP solution using padding. I can link to the commit once it is pushed (maybe in 20 minutes or so)

Reply:
This is my current solution which I still need to check in detail, but currently I think it is good https://github.com/OpenSenseAction/poligrain/blob/78ad8b3c7648c4cfae1f86ccb1b3b5d098379ab5/src/poligrain/spatial.py#L122-L136

Not sure if this is transferable to your code here.

Reply:
please note that there seems to be some bug in there...

Reply:
update: I just had a type in the notebook where I used and tested the code. Now I would say it works correctly.